core/session.ts

```diff
@@ -315,11 +315,24 @@ export class Session {
     return assertion;
   }
 
-  public declare(dataset: dataform.ITarget): Declaration {
-    const declaration = new Declaration();
-    declaration.session = this;
-    utils.setNameAndTarget(this, declaration.proto, dataset.name, dataset.schema, dataset.database);
-    declaration.proto.fileName = utils.getCallerFile(this.rootDir);
+  public declare(declarationConfig: dataform.ITarget | dataform.ActionConfig): Declaration {
+    // The declaration property exists on ActionConfig but not on Target.
+    if (!declarationConfig.hasOwnProperty("declaration")) {
```

Comment:
This is pretty nasty! Two paths here and two paths in the constructor... 
Surely there must be a way to push this all into the constructor at least?

Reply:
This has to be nasty somewhere in order to preserve the previous functionality - one alternative would be to just stop supporting Target as a parameter.

I'm not keen for it being in the constructor, because I think it obfuscates that it is the method on session that contains deprecated functionality, as opposed to the class constructor which isn't something the user can call directly.

Reply:
I still think there probably is some way to make this a little nicer, but I see you have a pattern here so let's stick to that until we see a better way :)